Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/task/AzureOperationEnhancer.java

```diff
@@ -20,45 +20,38 @@
  * SOFTWARE.
  */
 
-package com.microsoft.azure.toolkit.lib.common.operation;
+package com.microsoft.azure.toolkit.lib.common.task;
 
 import com.microsoft.azure.toolkit.lib.common.exception.AzureToolkitOperationException;
+import com.microsoft.azure.toolkit.lib.common.operation.AzureOperationRef;
 import lombok.extern.java.Log;
-import org.apache.commons.collections4.CollectionUtils;
 import org.aspectj.lang.JoinPoint;
 import org.aspectj.lang.annotation.*;
 import org.aspectj.lang.reflect.MethodSignature;
 
 import java.util.Objects;
-import java.util.logging.Level;
 
 @Aspect
 @Log
 public final class AzureOperationEnhancer {
 
-    @Pointcut("execution(@AzureOperation * *..*.*(..))")
+    @Pointcut("execution(@com.microsoft.azure.toolkit.lib.common.operation.AzureOperation * *..*.*(..))")
     public void operation() {
     }
 
     @Before("operation()")
     public void enterOperation(JoinPoint point) {
         final AzureOperationRef operation = toOperationRef(point);
-        final AzureOperation.Type type = AzureOperationUtils.getAnnotation(operation).type();
-        if (type == AzureOperation.Type.ACTION) {
-            AzureOperationsContext.clear();
-        }
-        AzureOperationsContext.push(operation);
+        log.info(String.format("enter operation[%s] in context[%s]", operation, AzureTaskContext.current()));
+        AzureTaskContext.current().pushOperation(operation);
     }
 
     @AfterReturning("operation()")
     public void exitOperation(JoinPoint point) {
         final AzureOperationRef operation = toOperationRef(point);
-        if (CollectionUtils.isNotEmpty(AzureOperationsContext.getOperations())) {
-            final AzureOperationRef popped = AzureOperationsContext.pop();
-            if (!Objects.equals(popped, operation)) {
-                log.log(Level.SEVERE, "!!!operation stack is wrongly managed!!!");
-            }
-        }
+        log.info(String.format("exit operation[%s] in context[%s]", operation, AzureTaskContext.current()));
+        final AzureOperationRef popped = AzureTaskContext.current().popOperation();
+        assert Objects.equals(popped, operation) : String.format("popped operation[%s] is not the exiting operation[%s]", popped, operation);
```

Comment:
should we break users operation when meet issues in operation tracking? I think `log.error` should be enough

Reply:
it's is a fatal error if this assertion fails. going forward will just cause unexpected errors.

Reply:
Operation annotation should not affect the logic, why will be fatal error here? Shall we catch exception here and log error?

Reply:
> Operation annotation should not affect the logic, why will be fatal error here?
IMO. the code should never come here, this assertion failure means our error handling framework has a error/bug, but not exception. and going forward is meaningless.
it's just similar to a framework's inner error or system error(e.g. `OutOfMemoryError`). which can not recover and the user can do nothing.

And we would expect the user report it to developer. 
As we have discussed, 
for `AzureToolkitException`/`AzureToolkitRuntimeException`, we will not show `report button`.
while for unknow exceptions(i.e. **non**-`AzureToolkitException`/`AzureToolkitRuntimeException`, including `AssertionException`/`NullPointerException`...), our root error handler will show a `report` button on the dialog/notification.